libs/template/renderer.go

```diff
@@ -0,0 +1,260 @@
+		}
+	}
+
+	templateFiler, err := filer.NewLocalClient(templateRoot)
+	if err != nil {
+		return nil, err
+	}
+
+	instanceFiler, err := filer.NewLocalClient(instanceRoot)
+	if err != nil {
+		return nil, err
+	}
+
+	ctx = log.NewContext(ctx, log.GetLogger(ctx).With("action", "initialize-template"))
+
+	return &renderer{
+		ctx:           ctx,
+		config:        config,
+		baseTemplate:  tmpl,
+		files:         make([]*inMemoryFile, 0),
+		skipPatterns:  make([]string, 0),
+		templateFiler: templateFiler,
+		instanceFiler: instanceFiler,
+	}, nil
+}
+
+// Executes the template by applying config on it. Returns the materialized template
+// as a string
+func (r *renderer) executeTemplate(templateDefinition string) (string, error) {
+	// Create copy of base template so as to not overwrite it
+	tmpl, err := r.baseTemplate.Clone()
+	if err != nil {
+		return "", err
+	}
+
+	// Parse the template text
+	tmpl, err = tmpl.Parse(templateDefinition)
+	if err != nil {
+		return "", err
+	}
+
+	// Execute template and get result
+	result := strings.Builder{}
+	err = tmpl.Execute(&result, r.config)
+	if err != nil {
+		return "", err
+	}
+	return result.String(), nil
+}
+
+func (r *renderer) computeFile(relPathTemplate string) (*inMemoryFile, error) {
+	// read template file contents
+	templateReader, err := r.templateFiler.Read(r.ctx, relPathTemplate)
+	if err != nil {
+		return nil, err
+	}
+	contentTemplate, err := io.ReadAll(templateReader)
+	if err != nil {
+		return nil, err
+	}
+
+	// execute the contents of the file as a template
+	content, err := r.executeTemplate(string(contentTemplate))
+	// Capture errors caused by the "fail" helper function
+	if target := (&ErrFail{}); errors.As(err, target) {
+		return nil, target
+	}
+	if err != nil {
+		return nil, fmt.Errorf("failed to compute file content for %s. %w", relPathTemplate, err)
+	}
+
+	// Execute relative path template to get materialized path for the file
+	relPath, err := r.executeTemplate(relPathTemplate)
+	if err != nil {
+		return nil, err
+	}
+
+	return &inMemoryFile{
+		path:    relPath,
+		content: []byte(content),
+	}, nil
+}
+
+func (r *renderer) walk() error {
+	directories := []string{"."}
+	var currentDirectory string
+
+	for len(directories) > 0 {
+		currentDirectory, directories = directories[0], directories[1:]
+
+		// Skip current directory if it matches any of accumulated skip patterns
+		instanceDirectory, err := r.executeTemplate(currentDirectory)
+		if err != nil {
+			return err
+		}
+		isSkipped, err := r.isSkipped(instanceDirectory)
+		if err != nil {
+			return err
+		}
+		if isSkipped {
+			logger.Infof(r.ctx, "skipping walking directory: %s", instanceDirectory)
+			continue
+		}
+
+		// Add skip function, which accumulates skip patterns relative to current
+		// directory
+		r.baseTemplate.Funcs(template.FuncMap{
+			"skip": func(relPattern string) string {
+				// patterns are specified relative to current directory of the file
+				// {{skip}} function is called from
```

Comment:
Unfinished comment?

Reply:
No, it's all a single comment, added some punctuation to make that clearer